File: leetcode/011_container_with_most_water.cpp
// Copyright (c) 2018 Jacopo Notarstefano

#include <algorithm>
#include <cassert>
#include <vector>

using namespace std;

class Solution {
 public:
    int maxArea(vector<int>& height) {
        int i = 0, j = height.size() - 1, result = 0;

        while (i < j) {
            result = max(result, min(height[i], height[j]) * (j - i));

            if (height[i] > height[j]) {
                j--;
            } else if (height[i] < height[j]) {
                i++;
            } else {
                i++;
                j--;
            }
        }

        return result;
    }
};

int main() {
    auto solution = Solution();

    vector<int> height = {5, 3, 4, 1, 2};

    assert(8 == solution.maxArea(height));
}

File: leetcode/011_container_with_most_water.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):
    def maxArea(self, height):
        i, j, result = 0, len(height) - 1, 0

        while i < j:
            result = max(result, min(height[i], height[j]) * (j - i))

            if height[i] < height[j]:
                i += 1
            elif height[i] > height[j]:
                j -= 1
            else:
                i += 1
                j -= 1

        return result


if __name__ == '__main__':
    solution = Solution()

    assert 49 == solution.maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])
